README.md
<p align="center">
<img src="https://ren-yc.github.io/assets/ClassTools/RRC.png" width="200">
</p>

<h1 align="center">- Random Roll Call -</h1>

<p align="center">
<img src="https://github.com/class-tools/RandomRollCall/actions/workflows/gitmsg.yml/badge.svg">
<img src="https://github.com/class-tools/RandomRollCall/actions/workflows/codeql.yml/badge.svg">
<img src="https://img.shields.io/github/v/release/class-tools/RandomRollCall.svg?logo=iCloud">
<img src="https://img.shields.io/github/downloads/class-tools/RandomRollCall/total?label=GitHub%20Downloads">
<img src="https://img.shields.io/badge/support-Windows-blue?logo=Windows">
<img src="https://img.shields.io/badge/support-Linux-blue?logo=Linux">
<img src="https://img.shields.io/badge/support-Mac%20OS-blue?logo=MacOS">
</p>

A small GUI random roll call program made by Python.

**English** | [简体中文](./README.zh-CN.md)

### Contributors

| <img src="https://avatars.githubusercontent.com/u/53416099?v=4" width="60px"></br> ren-yc (Yuchen Ren) |
| :---: |
| ![](https://shields.io/badge/Coding-green?logo=visual-studio-code&style=for-the-badge)<br>![](https://shields.io/badge/BugTester-yellow?logo=open-bug-bounty&style=for-the-badge) |

**For more information, please go to [Wiki](https://github.com/class-tools/RandomRollCall/wiki)**

CHANGELOG.md
# Changelog

This is the changelog of [Random Roll Call](https://github.com/class-tools/RandomRollCall).

这是 [随机点名](https://github.com/class-tools/RandomRollCall) 的更新日志。

This change log supports two languages: `en-US` `zh-CN`. (**Every two lines, the first line is `en-US` and the second line is `zh-CN`**)

这个更新日志支持两个语言：`en-US` `zh-CN`。（**每两行第一行为 `en-US` 第二行为 `zh-CN`**）

## 2.0.0

Replan the project.

重新规划项目。

Add CodeQL GitHub Action for static code scanning.

添加 CodeQL GitHub Action 以进行静态代码扫描。

Rename language identifier for Simplified Chinese.

重命名简体中文语言标识符。

Change data files format to Json (Because this module is built-in in Python).

更改数据文件格式为 Json （因为此模块内置在 Python 中）。

Change source code indent to Tab.

更改源代码缩进为 Tab。

When the setting file is missing, the file with the default configuration is automatically generated.

缺失设置文件时自动生成带有默认配置的文件。

The installation package is no longer planned to be provided.

计划不再提供安装包 （但我们允许其他人在发布的文件为基础制作并自由分发安装包）。

Update repository related files to the latest organization standards (However, we allow others to make and freely distribute installation packages based on published files).

将存储库相关文件更新至最新组织标准。

## 1.1.6.x

Add changing number. (The number changes continuously after the first click, and stops after the second click)

添加不断变化的数字。（第一次点击后数字不断变化，第二次点击后停止）

Add welcome text. ("Start")

添加欢迎文本。（“开始”）

## 1.1.5.x LTS

~~**This version of the program is long term supported.**~~

~~**此版本的程序被长期支持。**~~

Fix ISS #6.

修复 ISS #6。

Change Issues Checker Action runtime environment.

更改 Issues Checker Action 运行时环境。

Fix ISS #5. (Modify number range & The font size varies with the length of the number)

修复 ISS #5。（修改数字范围 & 字号随数字长度变化）

Add exception handling.

加入异常处理。

## 1.1.4.x LTS

~~**This version of the program is long term supported.**~~

~~**此版本的程序被长期支持。**~~

Add Github Action to delete issue of users without starring repository.

增加 Github Action 以删除没有 Star 存储库的用户的 Issue。

Add wiki. (Move information from readme to wiki)

增加 Wiki。（移动自述文件中的信息至 Wiki）

Update language yaml file.

更新语言 YAML 文件。

The modified minimum and maximum student numbers can be saved to the setting file.

修改最小及最大学号可以被保存至设置文件。

Increase the version number in the setting file.

在设置文件中增加版本号。

Add data file reading error display. (no longer unresponsive)

增加数据文件读取错误显示。（不再是无反应）

## 1.1.3.x

~~Add Github Action to check pull request code quality.~~ (Deleted in 1.1.4)

~~增加 Github Action 以分析拉取请求的代码质量。~~（在 1.1.4 中被删除）

~~Add Github Action to see result of code on Codecov _(Fixed three times)_.~~ (Deleted in 1.1.3.7)

~~增加 Github Action 以查看在 Codecov 上的代码分析结果 _（修复了三次）_~~（在 1.1.3.7 中被删除）

Edit language operation mode (using YAML).

修改语言运行模式（使用 YAML）。

Add settings file (using YAML).

增加设置文件（使用 YAML）。

**From now on, the program is not a single file (using the installation package).**

**从此版本起，本程序不再是单一编译文件（使用安装包）。**

## 1.1.2.x

Add "Help" to open readme in the browser.

增加 “帮助” 以在浏览器中打开自述文件。

Add dialog to ask the minimum and the maximum student number for random number.

加入对话框以询问最小及最大的学号以获取随机数。

Add changelog.

加入了更新日志。

Update readme.

更新自述文件。

## 1.1.0.x ~ 1.1.1.x

Add label to random roll call.

增加标签以获取随机数。

Add menubar.

增加菜单栏。

Update readme.

更新自述文件。

## Init & 1.0.0.x

Create Github Action to check commit message.

创建 Github Action 以检查提交信息格式。

Add MIT License.

添加 MIT 许可证。

Create readme.

创建自述文件。

RRC.py
'''
Random Roll Call Main Source File 2.0.0
This source code file is under MIT License.
Copyright (c) 2022 Class Tools Develop Team
Contributors: ren-yc
'''
# Modules
import os
import sys
import json
import atexit
import random
import tkinter
import webbrowser
from time import sleep
from tkinter.messagebox import showerror, showinfo
from tkinter.simpledialog import askinteger

# Variables
RUNNING = False
END = False
VERSION = '2.0.0'
TK_ROOT = tkinter.Tk()
SUPPORT_LANG = ['en-US', 'zh-CN']
BASE_DIR = (os.path.dirname(sys.executable) if hasattr(sys, 'frozen') else os.path.dirname(__file__))

# Error
def ERRNO1():
	showerror(title = 'Fatal Error', message = '[Errno 2] Cannot read language file data.')
	print('FATAL ERROR: [Errno 2] Cannot read language file data.')

def ERRNO2():
	showerror(title = 'Error', message = '[Errno 3] Number out of range. (0 < min or max < 100000 and min < max)')
	print('ERROR: [Errno 3] Number out of range. (0 < min or max < 100000 and min < max)')

# Read
def Read_Settings():
	global Settings_Dict
	Settings_Dict = {}
	try:
		with open(os.path.join(BASE_DIR, './settings.json'), 'r', encoding = 'utf-8') as f:
			Settings_Dict = json.load(f)
			assert Settings_Dict['Min'] > 0 and Settings_Dict['Max'] < 100000 and Settings_Dict['Min'] < Settings_Dict['Max']
	except (KeyError, AssertionError, FileNotFoundError, json.decoder.JSONDecodeError):
		Settings_Dict['Language'] = 'en-US'
		Settings_Dict['Min'] = 1
		Settings_Dict['Max'] = 49
		update()

def Read_Language():
	global Language_Dict
	try:
		with open(os.path.join(BASE_DIR, './lang/' + Settings_Dict['Language'] + '.json'), 'r', encoding = 'utf-8') as f:
			Language_Dict = json.load(f)
	except (FileNotFoundError, json.decoder.JSONDecodeError):
		ERRNO1()
		TK_ROOT.quit()
		sys.exit(1)

# Update Settings
def update():
	with open(os.path.join(BASE_DIR, './settings.json'), 'w', encoding = 'utf-8') as f:
		json.dump(Settings_Dict, f, indent = '\t')

# Dialog
def Ask_MIN():
	tmp = askinteger(title = Language_Dict['Ask_Min_Message'], prompt = Language_Dict['Ask_Min_Message'], initialvalue = Settings_Dict['Min'])
	if tmp is None:
		return
	if tmp <= 0 or tmp >= Settings_Dict['Max']:
		ERRNO2()
		return
	Settings_Dict['Min'] = tmp
	update()

def Ask_MAX():
	tmp = askinteger(title = Language_Dict['Ask_Max_Message'], prompt = Language_Dict['Ask_Max_Message'], initialvalue = Settings_Dict['Max'])
	if tmp is None:
		return
	if tmp >= 100000 or tmp <= Settings_Dict['Min']:
		ERRNO2()
		return
	Settings_Dict['Max'] = tmp
	update()

def FrontSize(tmp):
	length = len(str(tmp))
	if length <= 3:
		return 200
	elif length == 4:
		return 150
	elif length == 5:
		return 100
	else:
		return 50

# Menubar
def Show_Help():
	if Settings_Dict['Language'] == 'zh-CN':
		webbrowser.open('https://github.com/class-tools/RandomRollCall/blob/master/README.zh-CN.md')
	else:
		webbrowser.open('https://github.com/class-tools/RandomRollCall/blob/master/README.md')

def Show_About():
	showinfo(title = Language_Dict['About'], message = Language_Dict['About_Message1'] + VERSION + Language_Dict['About_Message2'] + Language_Dict['Version'])

def Switch_Lang(Lang: str):
	Settings_Dict['Language'] = Lang
	update()
	showinfo(title = Language_Dict['Warn'], message = Language_Dict['Reopen'])
	TK_ROOT.quit()

def Quit():
	global END
	END = True
	TK_ROOT.quit()
	atexit.unregister(Quit)

# Init
def Set_Window():
	global Language_Dict
	TK_ROOT.title(Language_Dict['Title'])
	TK_ROOT.geometry('400x400')
	TK_ROOT.resizable(width = False, height = False)
	TK_ROOT.iconphoto(False, tkinter.PhotoImage(file = os.path.join(BASE_DIR, './RRC.png')))

def Set_Menubar():
	global Language_Dict
	menubar = tkinter.Menu(TK_ROOT)
	TK_ROOT['menu'] = menubar
	settingsmenu = tkinter.Menu(menubar, tearoff = False)
	settingsmenu.add_command(label = Language_Dict['Ask_Min_Message'], command = Ask_MIN)
	settingsmenu.add_command(label = Language_Dict['Ask_Max_Message'], command = Ask_MAX)
	languagemenu = tkinter.Menu(settingsmenu, tearoff = False)
	for item in SUPPORT_LANG:
		languagemenu.add_radiobutton(label = item, state = 'disabled' if item == Settings_Dict['Language'] else 'normal', command = (lambda p = item: Switch_Lang(p)))
	settingsmenu.add_cascade(label = Language_Dict['Switch_Lang'], menu = languagemenu)
	menubar.add_cascade(label = Language_Dict['Settings'], menu = settingsmenu)
	menubar.add_command(label = Language_Dict['Help'], command = Show_Help)
	menubar.add_command(label = Language_Dict['About'], command = Show_About)
	menubar.add_command(label = Language_Dict['Quit'], command = Quit)

# Events
def click(events):
	global RUNNING
	global END
	if RUNNING == False:
		RUNNING = True
		while RUNNING == True and END == False:
			try:
				selected = random.randint(Settings_Dict['Min'], Settings_Dict['Max'])
				big_num_label.config(font = 'Helvetica -%d bold' % FrontSize(selected))
				big_num_label['text'] = selected
				big_num_label.update()
				sleep(0.01)
			except tkinter._tkinter.TclError:
				break
	else:
		RUNNING = False

# Mainloop
def main():
	global big_num_label
	big_num_label = tkinter.Label(TK_ROOT, text = Language_Dict['Window_Info'], width = 380, height = 380)
	big_num_label.config(font = 'Helvetica -%d bold' % 100)
	big_num_label.bind('<Button-1>', click)
	big_num_label.pack(side = tkinter.LEFT)
	TK_ROOT.mainloop()

if __name__ == '__main__':
	atexit.register(Quit)
	Read_Settings()
	Read_Language()
	Set_Window()
	Set_Menubar()
	main()


RandomRollCall.py
'''
RandomRollCall Main Source File 1.1.6
This source code file is under MIT License.
Copyright (c) 2022 Class Tools Develop Team
Contributors: ren-yc
'''
import webbrowser
import random
import yaml
import os
import sys
import tkinter as tk
from time import sleep
from tkinter.messagebox import showerror, showinfo
from tkinter.simpledialog import askinteger, askstring

# Init
root = tk.Tk()
supported_lang = ["en-Us", "zh-Hans"]
running = False

# Error
def errno_1():
    showerror(title = "Fatal Error", message = "[Errno 1] Cannot read settings file data.")
    print("FATAL ERROR: [Errno 1] Cannot read settings file data.")

def errno_2():
    showerror(title = "Fatal Error", message = "[Errno 2] Cannot read language file data.")
    print("FATAL ERROR: [Errno 2] Cannot read language file data.")

def errno_3():
    showerror(title = "Error", message = "[Errno 3] Number out of range. (0 < min or max < 100000 and min < max)")
    print("ERROR: [Errno 3] Number out of range. (0 < min or max < 100000 and min < max)")

# Yaml + Language
yamlSettingsPath = os.path.join("", "settings.yml")
try:
    SettingsFile = open(yamlSettingsPath, 'r', encoding = 'utf-8')
    dicSettings = yaml.load(SettingsFile.read(), Loader = yaml.FullLoader)
    default_Lang = dicSettings['Language']
    minnum = dicSettings['Min']
    maxnum = dicSettings['Max']
    if minnum == None or maxnum == None:
        raise TypeError
    version = dicSettings['Version']
    SettingsFile.close()
except:
    errno_1()
    root.quit()
    sys.exit(1)
try:
    yamlPath = os.path.join("", "lang/" + default_Lang + ".yml")
    file = open(yamlPath, 'r', encoding = 'utf-8')
    dic = yaml.load(file.read(), Loader = yaml.FullLoader)
    Lang_Title = dic['Title']
    Lang_Menu_Settings = dic['Settings']
    Lang_Menu_Help = dic['Help']
    Lang_Menu_About = dic['About']
    Lang_Menu_Quit = dic['Quit']
    Lang_Menu_MIN = dic['Ask_Min_Message']
    Lang_Menu_MAX = dic['Ask_Max_Message']
    Lang_Menu_Lang = dic['Switch_Lang']
    Lang_Switch = dic['Switch_Info']
    Lang_Messagebox_Message1 = dic['About_Message1']
    Lang_Messagebox_Message2 = dic['About_Message2']
    Lang_Warn = dic['Warn']
    Lang_Please_Reopen = dic['Reopen']
    Lang_Error = dic['Error']
    Lang_Switch_Error_Message = dic['Switch_Lang_Error']
    Lang_Window_Info = dic['Window_Info']
    Lang_Version = dic['Version']
    file.close()
except:
    errno_2()
    root.quit()
    sys.exit(1)

# Update Settings
def update():
    SettingsFile = open(yamlSettingsPath, 'w', encoding = 'utf-8')
    data = {"Language" : default_Lang, "Min" : minnum, "Max" : maxnum, "Version" : version}
    yaml.dump(data = data, stream = SettingsFile, allow_unicode = True, sort_keys = False)

# Dialog
def Ask_MIN():
    global minnum
    tmp = askinteger(title = Lang_Menu_MIN, prompt = Lang_Menu_MIN)
    if tmp == None:
        return
    if tmp <= 0 or tmp >= maxnum:
        errno_3()
        return
    minnum = tmp
    update()

def Ask_MAX():
    global maxnum
    tmp = askinteger(title = Lang_Menu_MAX, prompt = Lang_Menu_MAX)
    if tmp == None:
        return
    if tmp >= 100000 or tmp <= minnum:
        errno_3()
        return
    maxnum = tmp
    update()

def FrontSize(tmp):
    length = len(str(tmp))
    if length <= 3:
        return 200
    elif length == 4:
        return 150
    elif length == 5:
        return 100
    else:
        return 50

# Menubar
def Show_Help():
    if default_Lang == "zh-Hans":
        webbrowser.open("https://github.com/class-tools/RandomRollCall/blob/master/README.zh-Hans.md")
    else:
        webbrowser.open("https://github.com/class-tools/RandomRollCall/blob/master/README.md")

def Show_About():
    showinfo(title = Lang_Menu_About, message = Lang_Messagebox_Message1 + version + Lang_Messagebox_Message2 + Lang_Version)

def Ask_Lang():
    tmp = askstring(title = Lang_Menu_Lang, prompt = Lang_Switch)
    if tmp == None:
        return
    if not tmp in supported_lang:
        showerror(title = Lang_Error, message = Lang_Switch_Error_Message)
    else:
        global default_Lang
        default_Lang = tmp
        update()
        showinfo(title = Lang_Warn, message = Lang_Please_Reopen)
        root.quit()
        sys.exit(0)

menubar = tk.Menu(root)
root['menu'] = menubar
settingsmenu = tk.Menu(menubar, tearoff = False)
settingsmenu.add_command(label = Lang_Menu_MIN, command = Ask_MIN)
settingsmenu.add_command(label = Lang_Menu_MAX, command = Ask_MAX)
settingsmenu.add_command(label = Lang_Menu_Lang, command = Ask_Lang)
menubar.add_cascade(label = Lang_Menu_Settings, menu = settingsmenu)
menubar.add_command(label = Lang_Menu_Help, command = Show_Help)
menubar.add_command(label = Lang_Menu_About, command = Show_About)
menubar.add_command(label = Lang_Menu_Quit, command = root.quit)

# Window
root.title(Lang_Title)
root.geometry("400x400")
root.resizable(width = False, height = False)
root.iconbitmap("RandomRollCall.ico")

# Mainloop
def click(events):
    global running
    if running == False:
        running = True
        while running == True:
            selected = random.randint(minnum, maxnum)
            big_num_label.config(font = 'Helvetica -%d bold' % FrontSize(selected))
            big_num_label['text'] = selected
            big_num_label.update()
            sleep(0.01)
    else:
        running = False

# Show
big_num_label = tk.Label(root, text = Lang_Window_Info, width = 380, height = 380)
big_num_label.config(font = 'Helvetica -%d bold' % 100)
big_num_label.bind("<Button-1>", click)
big_num_label.pack(side = tk.LEFT)
root.mainloop()
